src/net/client/xfr.rs

```diff
@@ -0,0 +1,1147 @@
+//! Parsing of AXFR/IXFR response messages for higher level processing.
+//!
+//! This module provides [`XfrResponseProcessor`] which enables you to process
+//! one or more AXFR/IXFR response messages in terms of the high level
+//! [`XfrEvent`]s that they represent without having to deal with the
+//! AXFR/IXFR protocol details.
+use core::iter::Flatten;
+
+use std::fmt::Debug;
+
+use bytes::Bytes;
+use tracing::trace;
+
+use crate::base::iana::Opcode;
+use crate::base::message::AnyRecordIter;
+use crate::base::wire::ParseError;
+use crate::base::{Message, ParsedName, Record, Rtype, Serial};
+use crate::rdata::{AllRecordData, Soa};
+
+//------------ XfrRecord ------------------------------------------------------
+
+/// The type of record processed by [`XfrResponseProcessor`].
+pub type XfrRecord =
+    Record<ParsedName<Bytes>, AllRecordData<Bytes, ParsedName<Bytes>>>;
+
+//------------ XfrResponseProcessor -------------------------------------------
+
+/// An AXFR/IXFR response processor.
+///
+/// [`XfrResponseProcessor`] can be invoked on one or more sequentially
+/// AXFR/IXFR received response messages to verify them and during processing
+/// emit events which can be consumed via the iterator returned by
+/// [`process_answer()`].
+///
+/// Each [`XfrEventIterator`] produces events for a single response message.
+/// If the end of the XFR response sequence has been reached the iterator will
+/// emit an [`XfrEvent::TransferComplete`] event.
+///
+/// If the `TransferComplete` event has not been seen it means that the
+/// sequence is incomplete and the next response message in the sequence
+/// should be passed to [`process_next_answer()`] along with the exhausted
+/// iterator. This will populate thr [`XfrEventIterator`] with more records
+/// to parse thereby causing iteration to resume.
+///
+/// The process of producing and consuming iterators continues until the end
+/// of the transfer is detected or a parsing error occurs.
+pub struct XfrResponseProcessor;
```

Comment:
This doesn't seems to be object. It doesn't have any state. Maybe just put some functions in a module?

Reply:
This was an artefact of moving from callback to iterator approach. As the code has now been reverted to the callback approach this issue no longer exists.